Runtime/Python/src/python_bebop/bebop.py

```diff
@@ -0,0 +1,228 @@
+
+ticksBetweenEpochs = 621355968000000000
+dateMask = 0x3fffffffffffffff
+
+class UnionDefinition:
+    """
+    A utility class for unions
+    """
+    discriminator: int
+    value: Any
+
+    def __init__(self, discriminator: int, value: Any):
+        self.discriminator = discriminator
+        self.value = value
+
+UnionType = TypeVar("UnionType", bound=UnionDefinition)
+
+class BebopReader:
+    """
+    A wrapper around a bytearray for reading Bebop base types from it.
+
+    It is used by the code that `bebopc --lang python` generates. 
+    You shouldn't need to use it directly.
+    """
+
+    _emptyByteList = bytearray()
+    _emptyString = ""
+
+    def __init__(self, buffer=None):
+        self._buffer = buffer if buffer is not None else bytearray()
+        self.index = 0
+
+    @classmethod
+    def from_buffer(cls, buffer: bytearray):
+        return cls(buffer)
+
+    def _skip(self, amount: int):
+        self.index += amount
+
+    def read_byte(self):
+        byte = self._buffer[self.index]
+        self.index += 1
+        return byte
+
+    def read_uint16(self):
+        v = self._buffer[self.index : self.index + 2]
+        self.index += 2
+        return int.from_bytes(v, byteorder="little")
+
+    def read_int16(self):
+        v = self._buffer[self.index : self.index + 2]
+        self.index += 2
+        return int.from_bytes(v, byteorder="little", signed=True)
+
+    def read_uint32(self):
+        v = self._buffer[self.index : self.index + 4]
+        self.index += 4
+        return int.from_bytes(v, byteorder="little")
+
+    def read_int32(self):
+        v = self._buffer[self.index : self.index + 4]
+        self.index += 4
+        return int.from_bytes(v, byteorder="little", signed=True)
+
+    def read_uint64(self):
+        v = self._buffer[self.index : self.index + 8]
+        self.index += 8
+        return int.from_bytes(v, byteorder="little")
+
+    def read_int64(self):
+        v = self._buffer[self.index : self.index + 8]
+        self.index += 8
+        return int.from_bytes(v, byteorder="little", signed=True)
+
+    def read_float32(self):
+        v = self._buffer[self.index : self.index + 4]
+        self.index += 4
+        return unpack("<f", bytearray(v))[0]
+
+    def read_float64(self):
+        v = self._buffer[self.index : self.index + 8]
+        self.index += 8
+        return unpack("<d", bytearray(v))[0]
+
+    def read_bool(self):
+        return self.read_byte() != 0
+
+    def read_bytes(self):
+        length = self.read_uint32()
+        if length == 0:
+            return self._emptyByteList
+        v = self._buffer[self.index : self.index + length]
+        self.index += length
+        return v
+
+    def read_string(self):
+        length = self.read_uint32()
+        if length == 0:
+            return self._emptyString
+        v = self._buffer[self.index : self.index + length]
+        self.index += length
+        return "".join([chr(c) for c in v])
```

Comment:
It looks like you’re trying to convert the slice of bytes into a string by using a list comprehension that converts each byte into a character using the chr() function and then joins these characters into a string.

The inefficiency in this code comes from the fact that it's converting each byte to a character one by one and then joining them together, which is not necessary. You’re already using a built-in method to encode the string to bytes, and the builtin decode is faster.


Try:
```

def read_string(self):
    length = self.read_uint32()
    if length == 0:
        return self._emptyString
    string_data = self._buffer[self.index : self.index + length]
    self.index += length
    return string_data.decode('utf-8')
```

Reply:
Oh thanks!